### PR TITLE
Cosmos (ATOM) integration

### LINKED ISSUE #92: Support Cosmos transactions

### PR DESCRIPTION
**Pull request by jamesscaur**
(Hopefully) closes #92

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -1,11 +1,12 @@
 package main
 
 import (
+	"strings"
```
Comment (terorie) [inline config.go]: Pointless change

File: platform/cosmos/api_test.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+			  "value": "A21fdP6IbVC9hER5smiim8I4EbFeIF/bW81IKwmmsdjH"
+			},
+			"signature": "MuR85p714L94tCenogRqzLh1bsbmhKTjs1L9JJPdhSVwQKh61EGlLqYGoUeN/n9xb+OOR9ESUOh2CAzVulKoVQ=="
+		  }
+		],
+		"memo": ""
+	  }
+	},
+	"timestamp": "2019-05-04T17:57:57Z"
+  }
```
Comment (terorie) [inline platform/cosmos/api_test.go]: Wrong indent

File: platform/cosmos/client.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+		url.Values{
+			"recipient": {address},
+			"page":      {strconv.FormatInt(1, 10)},
+			"limit":     {strconv.FormatInt(models.TxPerPage, 10)},
+		}.Encode())
+
+	fmt.Println(uri)
+
+	res, err := c.HTTPClient.Get(uri)
+	if err != nil {
```
Comment (terorie) [inline platform/cosmos/client.go]: Missing return

File: platform/cosmos/client.go [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+func (c *Client) GetOutputs(address string) (txs []Tx, err error) {
+	uri := fmt.Sprintf("%s/txs?%s",
+		c.BaseURL,
+		url.Values{
+			"sender": {address},
+			"page":   {strconv.FormatInt(1, 10)},
+			"limit":  {strconv.FormatInt(models.TxPerPage, 10)},
+		}.Encode())
+
+	fmt.Println(uri)
```
Comment (terorie) [inline platform/cosmos/client.go]: Useless print

File: platform/cosmos/client.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+	res, err := c.HTTPClient.Get(uri)
+	if err != nil {
```
Comment (terorie) [inline platform/cosmos/client.go]: Missing return

File: platform/cosmos/client.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+		logrus.WithError(err).Errorf("Cosmos: Failed to get transactions for address %s", address)
+	}
+
+	dec := json.NewDecoder(res.Body)
+	err = dec.Decode(&txs)
+	return
+}
+
+// GetOutputs - get all ATOM transactions for a given address, via sender
+func (c *Client) GetOutputs(address string) (txs []Tx, err error) {
```
Comment (terorie) [inline platform/cosmos/client.go]: Looks like this function does exactly the same as `GetInputs`

File: test/main.go [hunk 7]
```diff
@@ -40,8 +40,9 @@ var addresses = map[string]Entry{
 	"thundertoken":     {coin.TT,  "0x0ad80a408eac4f17ba0a9de8a12d8736f60700c3"},
 	"icon":             {coin.ICX, "hxee691e7bccc4eb11fee922896e9f51490e62b12e"},
 	"tron":             {coin.TRX, "TMuA6YqfCeX8EhbfYEg5y7S4DqzSJireY9"},
-	"vechain":          {coin.VET, "0xB5e883349e68aB59307d1604555AC890fAC47128"},
+	"vechain":          {coin.VET,  "0xB5e883349e68aB59307d1604555AC890fAC47128"},
```
Comment (terorie) [inline test/main.go]: Unrelated change

File: README.md [hunk 8]
```diff
@@ -23,6 +23,7 @@ __Supported Coins__
 <a href="https://icon.foundation" target="_blank"><img src="https://raw.githubusercontent.com/TrustWallet/tokens/master/coins/74.png" width="32" /></a>
 <a href="https://tron.network" target="_blank"><img src="https://raw.githubusercontent.com/TrustWallet/tokens/master/coins/195.png" width="32" /></a>
 <a href="https://vechain.org/" target="_blank"><img src="https://raw.githubusercontent.com/TrustWallet/tokens/master/coins/818.png" width="32" /></a>
+<a href="https://cosmos.network/" target="_blank"><img src="https://raw.githubusercontent.com/TrustWallet/tokens/master/coins/118.png" width="32" /></a>
```
Comment (terorie) [inline README.md]: Mind moving that to the end? Coins are usually ordered by date integrated

File: coin/coins.json [hunk 9]
```diff
@@ -307,5 +307,12 @@
 		"symbol": "WAVES",
 		"name": "Waves",
 		"link": "https://wavesplatform.com/"
+	},
+	{
+		"index": 118,
+		"hex": "0x80000076",
+		"symbol": "ATOM",
+		"name": "Cosmos ATOM",
+		"link": "https://cosmos.network/"
```
Comment (terorie) [inline coin/coins.json]: Needs number of decimals

File: platform/cosmos/model.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+// TxData - "tx" sub object
+type TxData struct {
+	TxType     string     `json:"type"`
+	TxContents TxContents `json:"value"`
+}
+
+// TxContents - amount, fee, and memo
+type TxContents struct {
+	TxMessage []TxMessage `json:"msg"`
```
Comment (terorie) [inline platform/cosmos/model.go]: No need to prefix everything with `Tx`, `Message`, `Fee`, `Memo` is fine

File: coin/slip44.go [hunk 11]
```diff
@@ -117,6 +121,12 @@ var Coins = map[uint]Coin {
 		Title: "ARK",
 		Website: "https://ark.io",
 	},
+	XMR: {
+		Index: XMR,
+		Symbol: "XMR",
+		Title: "Monero",
```
Comment (vikmeup) [inline coin/slip44.go]: why did you add XMR and QTUM?

File: platform/cosmos/model.go [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	TxAmount []Amount `json:"amount"`
+}
+
+// TxFee - also references the "amount" struct
+type TxFee struct {
+	FeeAmount []Amount `json:"amount"`
+}
+
+// Amount - the asset & quantity. Always seems to be enclosed in an array/list for some reason.
+// Perhaps used for multiple tokens transferred in a single sender/reciever transfer?
```
Comment (terorie) [inline platform/cosmos/model.go]: Can you take a look in the Cosmos docs? 90% sure though that it's just going to be one item

File: coin/slip44.go (hunk 11 above)
Comment (terorie) [reply coin/slip44.go]: That was auto-generated, my fault

Comment (jamesscaur) [reply coin/slip44.go]: Will remove it in update

File: config.go (hunk 1 above)
Comment (jamesscaur) [reply config.go]: Don't know why that happened, will remove

File: platform/cosmos/api_test.go (hunk 2 above)
Comment (jamesscaur) [reply platform/cosmos/api_test.go]: What indent should be used?

File: config.go (hunk 1 above)
Comment (jamesscaur) [reply config.go]: Hm, I can't seem to save my file without it appearing at the top. Perhaps it's something to do with a VS Code extension I have installed - doing local imports before Github imports... either way, I'll remove the change

File: platform/cosmos/client.go (hunk 3 above)
Comment (jamesscaur) [reply platform/cosmos/client.go]: Which line?

File: platform/cosmos/client.go (hunk 5 above)
Comment (jamesscaur) [reply platform/cosmos/client.go]: My bad, fixed

File: platform/cosmos/client.go (hunk 6 above)
Comment (jamesscaur) [reply platform/cosmos/client.go]: Good point, it does but it uses "sender" as a request parameter instead of "recipient". I can improve it so that "input" or "output" is included as a variable in the function and it chooses based on that.

File: test/main.go (hunk 7 above)
Comment (jamesscaur) [reply test/main.go]: Yep will leave out

File: README.md (hunk 8 above)
Comment (jamesscaur) [reply README.md]: Yep will do

File: coin/coins.json (hunk 9 above)
Comment (jamesscaur) [reply coin/coins.json]: I'll find this out - this is X where (sats ÷ (10^X)) = value, right?

File: platform/cosmos/model.go (hunk 10 above)
Comment (jamesscaur) [reply platform/cosmos/model.go]: Will remove, cheers

File: platform/cosmos/model.go (hunk 12 above)
Comment (jamesscaur) [reply platform/cosmos/model.go]: Will do

Comment (jamesscaur) [reply platform/cosmos/model.go]: I couldn't find anything in the docs about it - shall I hardcode it to expect only one array item?

**Comment by jamesscaur** [comment]
Rebasing, will squash & improve the commit messages

File: platform/cosmos/api_test.go (hunk 2 above)
Comment (terorie) [reply platform/cosmos/api_test.go]: That's up to you, but it's a mix of tabs and spaces right now

**Comment by jamesscaur** [comment]
Have rebased, can't figure out how to upload without having to download my (now squashed) commits - will create a new PR

**Comment by terorie** [comment]
@jamesscaur You can also force push instead of creating new PRs 😅

**Comment by jamesscaur** [comment]
> @jamesscaur You can also force push instead of creating new PRs 😅

Ah, thanks - will remember for next time.
